### ISSUE TITLE
Additional logic in determing access level

### ISSUE THREAD

**Issue by jasonpartyka** [palantirnet/workbench_access#122]
# Problem
Sometimes the ability to update/create content that is controlled by a term needs to be restricted further. As an example: 

- A term representing group "A" exists, but due to business reasons new content may not be assigned to this group.  The term cannot be removed as the association needs to be kept intact because of reporting, theme, et al. Let's call this the 'deprecated' group.
- Content part of "group A" may still be updated, with the deprecated group intact. However, only users who are members of the deprecated group may update it. The content may still in group "A", but if ownership is changed, it must go to a group that is not deprecated.

# Proposed Solution
In the generic case, provide a way that a developer can write additional logic to further restrict if access is denied. 

There are two ways we could go about this, and both involve updating `Drupal\workbench_access\Plugin\EntityReferenceSelection\TaxonomyHierarchySelection::getReferenceableEntities` method to execute the chosen pattern.

1. Create a service class to provide the logic. The developer then writes a service class do to provide the logic.
2. Provide an 'empty method' (for lack of a better term) that be used in a Plugin class that inherits from `TaxonomyHierarchySelection` and then implements this method. This 'empty method' would simply always say access is allowed in the default implementation within  `TaxonomyHierarchySelection`.

## Other solutions
These other solutions have been considered and have their own drawbacks:

1. Validate on save. This would preserve data integrity, but presents a sub-optimal user experience as the user may choose an option that is believed to be valid.
2. Completely override Plugin w/o inheriting from class. This is possible, but requires keeping an eye on the copied class instead of leveraging OO patterns.

I will need to make this alteration for our own needs, so I'd like to do it in a sustainable way that could be incorporated back into the module.

**Comment by larowlan** [comment]
Option 3: use the plugin alter hook to change the class used for the
taxonomy plugin, subclass the exisiting one and override get referenced
entities, calling the parent in most cases?

**Comment by agentrickard** [comment]
It's probably the plugin alter. Note that the class @jasonpartyka is mentioning here just controls the _autocomplete_ selection of terms. We're really looking back a step at the `alterForm()` method for the plugin class.

`Drupal\workbench_access\Plugin\AccessControlHierarchy\Taxonomy\AlterForm`

**Comment by larowlan** [comment]
Both are plugins that can be altered though

**Comment by jasonpartyka** [comment]
So, `getReferenceableEntities` is only used to determine autocomplete and is not called for form validation? If that's correct then the proposed solution is incomplete.

**Comment by larowlan** [comment]
Form validation should be from a validation plugin, so it works for REST et
al

**Comment by jasonpartyka** [comment]
Yeah, so then why not create a service so that the custom access logic is written just once? But, I don't see a validation plugin in workbench access either. Did I miss it? I'm looking at `8.x-1.0-beta2`

**Comment by larowlan** [comment]
Core uses selection plugins for validating as well as showing options

See valid referenced entity validation plugin in core.

**Comment by jasonpartyka** [comment]
Gotcha. So I'm totally overthinking it then. :-)

**Comment by agentrickard** [comment]
All that said, I'm not certain that we handle Validation correctly via plugins -- unless @larowlan is saying that core already handles that for us.

This is an area of the code where changes from D7 to D8 have me a bit behind.

**Comment by larowlan** [comment]
I think core does it for us *if* you select the right selection plugin for
the ER field

**Comment by agentrickard** [comment]
I believe we do in the case of the autocomplete -- because in that case we swap out the selection plugin in its entirety. I'm not entirely certain that the standard form alter that we do for other field types adds this validation step.

**Comment by jasonpartyka** [comment]
So, I don't see any Validation plugins in in the [src/Plugin](https://github.com/agentrickard/workbench_access/tree/8.x-1.x/src/Plugin) directory. This would handle any widget used for term selection, correct?

**Comment by jasonpartyka** [comment]
So -- I'm picking this issue up again. @agentrickard and @larowlan, do you believe there is any value in providing a generic solution?

**Comment by agentrickard** [comment]
Very probably, but I haven't given it much thought.
